typing/typemod.ml

```diff
@@ -1449,262 +1540,299 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
+                   in
+                   (id, name, mty, modl, mty', attrs, loc))
+                decls sbind in
+            let newenv = (* allow aliasing recursive modules from outside *)
+              List.fold_left
+                (fun env md ->
+                   let mdecl =
+                     {
+                       md_type = md.md_type.mty_type;
+                       md_attributes = md.md_attributes;
+                       md_loc = md.md_loc;
+                     }
+                   in
+                   Env.add_module_declaration ~check:true md.md_id mdecl env
+                )
+                env decls
+            in
+            let str_rem, sig_rem, final_env = type_struct newenv srem in
+            let bindings2 =
+              check_recmodule_inclusion newenv bindings1 in
+            mkstr (Tstr_recmodule bindings2) newenv loc :: str_rem,
+            (map_rec (fun rs mb ->
+                Sig_module(mb.mb_id, {
+                    md_type=mb.mb_expr.mod_type;
+                    md_attributes=mb.mb_attributes;
+                    md_loc=mb.mb_loc;
+                  }, rs))
+               bindings2 []) @ sig_rem,
+           final_env 
+        | Pstr_modtype pmtd ->
+            (* check that it is non-abstract *)
+            let newenv, mtd, sg =
+              transl_modtype_decl names env pmtd
+            in
+            let str_rem, sig_rem, final_env = type_struct newenv srem in
+            mkstr (Tstr_modtype mtd) newenv loc :: str_rem, sg :: sig_rem,
+            final_env
+        | Pstr_open sod -> begin
+            let (inserted_md, od) = type_open ~toplevel env sod in
+            let open_env = od.open_env in
+            let str_rem, sig_rem, final_env = type_struct open_env srem in
+            match inserted_md with
+            | None -> mkstr (Tstr_open od) open_env loc :: str_rem,
+                      sig_rem, final_env
+            | Some (id, md, md_env) ->
+                let tm =
+                  Tstr_module {mb_id=id;
+                               mb_name={txt=Ident.name id; loc=Location.none};
+                               mb_expr = od.open_expr;
+                               mb_attributes=od.open_expr.mod_attributes;
+                               mb_loc=od.open_expr.mod_loc} in
+                let tm_str = { str_desc = tm; str_loc = loc; str_env = md_env } in
+                let open_str = mkstr (Tstr_open od) open_env loc in
+                let md_sig =
+                  Sig_module (id, {md_type=md.Types.md_type; md_loc=loc;
+                                   md_attributes = []}, Trec_not) in
+                if not (in_nested_struct ()) then
+                  let s_rem = Mty_signature sig_rem in
+                  match Mtype.nondep_supertype open_env id s_rem with
+                  | Mty_signature sg ->
+                      tm_str :: open_str :: str_rem, md_sig :: sg, final_env
```

Comment:
Do we actually need/want to add the Sig_module here?  It will be removed by simplify_signature below (but only for implementation structures, not nested one).

Reply:
This is necessary for nested open. 

Consider this example

```OCaml
open struct
  open struct let counter = ref 0 end
  let inc () = incr counter
end
let () = inc()
```

which can be equivalent to
```OCaml
module M#0 = struct
  module M#1 = struct
    let counter = ref 0
  end
  open M#1
  let inc () = incr counter
end
open M#0
let () = inc ()
```

If we don't insert `Sig_module` for `M#1` in `M#0`, when we later `open M#0`, `inc` will have position 0 with respect to module `M#0`. This is not true, since position 0 in module `M#0` is module `M#1` and position 1 in module `M#0` will be `inc`. At the very end, you will get a segfault.

Reply:
Would that be addressed by https://github.com/ocaml/ocaml/pull/1610?